### PR TITLE
PIM-6856: List family variant on family edit page

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

Add the family variant list page

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | NA
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Done
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/DataGridBundle/EventListener/AddParametersToGridListener.php [hunk 1]
```diff
@@ -40,11 +40,12 @@ class AddParametersToGridListener
      * @param RequestParameters $requestParams Request params
      * @param bool              $isEditMode    Whether or not to add data_in, data_not_in params to query
      */
-    public function __construct($paramNames, RequestParameters $requestParams, $isEditMode = false)
+    public function __construct($paramNames, RequestParameters $requestParams, $isEditMode = false, $isQueryParam = false)
     {
         $this->paramNames = $paramNames;
         $this->requestParams = $requestParams;
         $this->isEditMode = $isEditMode;
+        $this->isQueryParam = $isQueryParam;
```
Comment (damien-carcel) [inline src/Pim/Bundle/DataGridBundle/EventListener/AddParametersToGridListener.php]: Please declare the new variable.

File: src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FamilyVariantDatasource extends RepositoryDatasource
+{
+    /** @var NormalizerInterface */
+    protected $normalizer;
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php]: private

File: src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    public function getResults()
+    {
+        $familyVariants = $this->qb->getQuery()->execute();
+
+        return array_map(function ($familyVariant) {
+            return new ResultRecord(
+                $this->normalizer->normalize(
+                    $familyVariant,
+                    'datagrid',
+                    ['localeCode' => $this->getParameters()[':localeCode']]
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php]: What happens if ':localeCode' does not exist?

File: src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Pim\Bundle\DataGridBundle\Datasource;
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php]: I think you should declare strict type for new classes

File: src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+use Oro\Bundle\DataGridBundle\Datasource\ResultRecord;
+use Pim\Bundle\DataGridBundle\Datasource\ResultRecord\HydratorInterface;
+use Pim\Bundle\DataGridBundle\Doctrine\ORM\Repository\DatagridRepositoryInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php]: extra line

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
+<?php
+
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: strict type

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalize family variant for the family variant grid
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class FamilyVariantNormalizer extends ProductNormalizer
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: extends ProductNormalizer, I guess this is wrong!

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            'familyCode'        => $familyVariant->getFamily()->getCode(),
+            'familyVariantCode' => $familyVariant->getCode(),
+            'label'             => isset($labels[$context['localeCode']]) ? $labels[$context['localeCode']] : $familyVariant->getCode(),
+            'level_1'           => '',
+            'level_2'           => ''
+        ];
+
+        foreach($familyVariant->getVariantAttributeSets() as $attributeSet) {
+            $axesCodes = array_map(function ($attribute) {
+                return $attribute->getLabel();
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: the variable is called `axesCodes` but you get the labels?

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php (hunk 6 above)
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: please add `declare(strict_types=1);`

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\EntityRepository;
+use Pim\Bundle\DataGridBundle\Doctrine\ORM\Repository\DatagridRepositoryInterface;
+
+/**
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FamilyVariantRepository extends EntityRepository implements DatagridRepositoryInterface
```
Comment (aRn0D) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php]: you don't need to extends `EntityRepository`, inject the entity manager in that class

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<?php
+
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php]: please add `declare(strict_types=1);`

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        parent::__construct($em, $em->getClassMetadata($class));
+    }
+
+    /**
+     * @param array $parameters
+     *
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    public function createDatagridQueryBuilder($parameters = [])
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php]: missing type

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family/form/family-variant.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            /**
+             * {@inheritdoc}
+             */
+            render: function () {
+                if (!this.variantGrid) {
+                    this.variantGrid = new Grid(
+                        'family-variant-grid',
+                        {
+                            family_id: this.getFormData().meta.id,
+                            localeCode: 'en_US'
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/family/form/family-variant.js]: Seems weird to hardcode the locale no?

**Comment by aRn0D** [comment]
Don't forget the change log

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family/form/family-variant.js (hunk 12 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/family/form/family-variant.js]: Oh, yeah, mistake :/

File: src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php (hunk 4 above)
Comment (juliensnz) [reply src/Pim/Bundle/DataGridBundle/Datasource/FamilyVariantDatasource.php]: done

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php (hunk 6 above)
Comment (juliensnz) [reply src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: done

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php (hunk 10 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilyVariantRepository.php]: done

File: src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php (hunk 6 above)
Comment (juliensnz) [reply src/Pim/Bundle/DataGridBundle/Normalizer/FamilyVariantNormalizer.php]: done

File: src/Pim/Bundle/DataGridBundle/spec/Normalizer/FamilyVariantNormalizerSpec.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+class FamilyVariantNormalizerSpec extends ObjectBehavior
+{
+    function let(NormalizerInterface $translationNormalizer)
+    {
+        $this->beConstructedWith($translationNormalizer);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Pim\Bundle\DataGridBundle\Normalizer\FamilyVariantNormalizer');
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/spec/Normalizer/FamilyVariantNormalizerSpec.php]: FamilyVariantNormalizer::class

File: src/Pim/Bundle/DataGridBundle/spec/Normalizer/FamilyVariantNormalizerSpec.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    function it_is_a_normalizer()
+    {
+        $this->shouldImplement('Symfony\Component\Serializer\Normalizer\NormalizerInterface');
```
Comment (aRn0D) [inline src/Pim/Bundle/DataGridBundle/spec/Normalizer/FamilyVariantNormalizerSpec.php]: idem

**Pull request merged**